### PR TITLE
[Explainability Evaluation] - GNN model (Un)Faithfulness v2

### PR DESCRIPTION
**Pull request by ZeynepP**
Attempt for the implementation of faithfulness measure according to [Probing GNN Explainers: A Rigorous Theoretical and Empirical Analysis of GNN Explanation Methods](https://arxiv.org/abs/2106.09078)

### CODE REVIEW

File: torch_geometric/explain/metric/faithfulness.py [hunk 1]
```diff
@@ -71,3 +71,84 @@ def unfaithfulness(
 
     kl_div = F.kl_div(y.log(), y_hat, reduction='batchmean')
     return 1 - float(torch.exp(-kl_div))
+
+
+# Probing GNN Explainers: A Rigorous Theoretical and Empirical Analysis of GNN Explanation Methods
+
+
+# code for rewiring edges, should be an input param rewiring_edges = True False in faitfulness function
```
Comment (amirtha255) [inline torch_geometric/explain/metric/faithfulness.py:79]: Typo: faitfulness to faithfulness

File: torch_geometric/explain/metric/faithfulness.py [hunk 2]
```diff
@@ -71,3 +71,84 @@ def unfaithfulness(
...
+
+    # getting subgraph
+    subgraph = explanation.get_explanation_subgraph()
+    if threshold_config is not None:
+        subgraph = subgraph.threshold(threshold_config)
+
+    node_mask, edge_mask = subgraph.node_mask, subgraph.edge_mask
+
+    # getting perturbated_nodes
+    # TODO : add a function with diffent option of perturbasion uniform,gaussian etc.
```
Comment (amirtha255) [inline torch_geometric/explain/metric/faithfulness.py:111]: Type: perturbasion : Perturbation

File: torch_geometric/explain/metric/faithfulness.py [hunk 3]
```diff
@@ -71,3 +71,84 @@ def unfaithfulness(
...
+def faithfulness(explainer: Explainer, explanation: Explanation,
+                 threshold_config: ThresholdConfig = None, num_iterations=10,
+                 rewirings_edges=True, **kwargs):
+    r"""Evaluates how faithful an :class:`~torch_geometric.explain.Explanation`
+       is to an underyling GNN predictor, as described in the
+       `"Probing GNN Explainers: A Rigorous Theoretical and Empirical Analysis of GNN Explanation Methods"
+       <https://arxiv.org/abs/2106.09078>`_ paper.
+    """
+    # TODO: all this to a config or to params
+    degree = 3
```
Comment (amirtha255) [inline torch_geometric/explain/metric/faithfulness.py:99]: Could add a comment about why it is of degree 3

File: torch_geometric/explain/metric/faithfulness.py [hunk 4]
```diff
@@ -71,3 +71,84 @@ def unfaithfulness(
...
+        subgraph = subgraph.threshold(threshold_config)
+
+    node_mask, edge_mask = subgraph.node_mask, subgraph.edge_mask
+
+    # getting perturbated_nodes
+    # TODO : add a function with diffent option of perturbasion uniform,gaussian etc.
+    sub_x = subgraph.x
+    perturbed_nodes = [sub_x.clone()]
+    for i in range(num_iterations):
+        cont_noise = np.random.normal(loc=pert_loc, scale=pert_scale,
```
Comment (amirtha255) [inline torch_geometric/explain/metric/faithfulness.py:115]: Why is cont_noise in a loop if loop iterator is not used ?
